docs/RFCS/table_descriptor_lease.md

```diff
@@ -0,0 +1,309 @@
+held by its previous incarnation. This will ensure that a transient
+failure does not require the lease expiration to occur before allowing
+a schema modification. When a node dies permanently or becomes
+unresponsive (e.g. detached from the network) the lease master will
+have to wait for any leases that node held to expire. This will result
+in an inability to perform more than one schema modification step to
+the descriptors referred to by those leases. The maximum amount of
+time this condition can exist is the lease duration.
+
+If the lease master dies a new master will be "elected" within
+20s. Lease state will be stored in a `leaseTable` that will be read
+when the master starts:
+
+```sql
+CREATE TABLE leaseTable (
+  DescID     INT,
+  Version    INT,
+  NodeID     INT,
+  Expiration TIMESTAMP,
+  PRIMARY KEY (DescID, Version, NodeID)
+)
+```
+
+Entries in the lease table will be added and removed as leases are
+acquired and released. A background goroutine will periodically delete
+expired leases. The master will maintain the invariant that for a
+given descriptor there are only two active versions (i.e. non-expired
+versions) in the lease table and that the two versions correspond to
+the table descriptor in the KV store and its immediate predecessor.
+
+When the lease master is unavailable (because it has died and a new
+master has not yet started), new leases cannot be granted but existing
+leases remain valid until they expire.
+
+# Drawbacks
+
+* This is complex. Perhaps overly complex. Where can it be simplified?
+
+* The watch mechanism is described with the assumption of streaming
+  responses. Unless we adopt gRPC sooner, we'll have to fallback to
+  some sort of cursor mechanism.
+
+# Alternatives
+
+* The master could avoid storing the lease state, but we would have to
+  wait for the lease duration at master startup before allowing any
+  schema modifications. There might be some hacks around this. For
+  example, if the master starts up and it is the first master (because
+  no `/lease-master` key existed) then it can avoid
+  waiting. Alternately, the master can persist only the largest lease
+  expiration and wait until that expiration passes before allowing
+  schema modifications. On the other hand, persisting the lease state
+  allows inspection of it via the SQL console.
+
+* Keeping track of local references to descriptor versions in order to
+  early release leases adds complexity. We could just wait for leases
+  to expire, though that would cause a 3-step schema modification to
+  take at least 10m to complete.
+
+# Unresolved questions
+
+* We need a mechanism to enforce that write operations do not commit
+  if they reference a table descriptor lease which has expired. The F1
+  paper talks about a "write fence" mechanism in spanner which allows
+  a transaction to specify a time before which the transaction needs
+  to commit otherwise it should be aborted.
```

Comment:
this looks like we could implement it rather easily: add an optional MaxTimestamp field which, if set, is used by EndTransaction. If proposed commit timestamp is ahead, instead abort the transaction (of course this can be cancelled earlier already).